src/main/java/com/ably/kafka/connect/client/DefaultAblyBatchClient.java

```diff
@@ -52,29 +67,118 @@ public DefaultAblyBatchClient(ChannelSinkConnectorConfig connectorConfig, Channe
                 logger.error("Error while sending batch", e);
+                sendMessagesToDLQ(records, dlqReporter);
+            }
+        }
+    }
+
+    /**
+     * Function to handle partial failure
+     * @param response Ably response
+     * @param records  Sink records
+     * @param dlqReporter DLQ reporter
+     */
+    public void handlePartialFailure(HttpPaginatedResponse response,
+                                     Map<String, List<SinkRecord>> records,
+                                     ErrantRecordReporter dlqReporter) {
+
+        Set<String> failedChannels = new HashSet<>();
+        JsonElement[] elements = response.items();
+        for (JsonElement element : elements) {
+            failedChannels.addAll(getFailedChannels(element));
+        }
+
+        for(String channel : failedChannels) {
+            List<SinkRecord> failedRecords = records.get(channel);
+            if(failedRecords != null || !failedRecords.isEmpty()) {
+                sendMessagesToDLQ(failedRecords, dlqReporter);
+            }
+        }
+    }
+
+    /**
+     * Function to parse Ably Batch API response
+     * @param response Ably response
+     * @return ABLY_BATCH_RESPONSE
+     */
+    public ABLY_BATCH_RESPONSE parseAblyBatchAPIResponse(HttpPaginatedResponse response) {
+
+        ABLY_BATCH_RESPONSE ablyBatchResponse = ABLY_BATCH_RESPONSE.SUCCESS;
+
+        String statusCode = Integer.toString(response.statusCode);
+        if(statusCode.startsWith(ABLY_REST_API_ERROR_CODE_4XX) || statusCode.startsWith(ABLY_REST_API_ERROR_CODE_5XX)) {
+            // DLQ all the sink records.
+            logger.error("Ably Batch API call failed with error code: " + statusCode);
+            ablyBatchResponse = ABLY_BATCH_RESPONSE.FAILURE;
+        } else if(statusCode.startsWith(ABLY_REST_API_ERROR_CODE_2XX)) {
+            JsonElement[] elements = response.items();
+            for (JsonElement element : elements) {
+                int failureCount = element.getAsJsonObject().getAsJsonPrimitive("failureCount").getAsInt();
+                if (failureCount > 0) {
+                    // DLQ only the failed sink records for that specific channel.
+                    logger.error("Ably Batch API call failed with error code: " + statusCode);
+                    ablyBatchResponse = ABLY_BATCH_RESPONSE.PARTIAL_FAILURE;
+                    break;
+                }
             }
         }
+
+        return ablyBatchResponse;
+    }
+
+    /**
+     * Function to send sink records to DLQ.
+     * @param records
+     * @param dlqReporter
+     */
+    private void sendMessagesToDLQ(List<SinkRecord> records,
+                                   ErrantRecordReporter dlqReporter) {
+        if(dlqReporter == null) {
+            logger.error("Dead letter queue not configured");
+            return;
+        }
+        for(SinkRecord record: records) {
+            dlqReporter.report(record, new Throwable("Ably Batch call failed"));
+        }
```

Comment:
We should really be attaching the specific error that came back from Ably here, rather than a generic "batch call failed" message. If the whole batch failed, there should be a status code and message for the whole response. If it's a partial failure, the error code and message for the corresponding channel should be attached. Otherwise users will have no idea why messages are sent to the DLQ and will be forced to reproduce the problem before they can take action.

Reply:
Added logic for partial failure, not sure what the response format of a whole batch failure looks like, it will be good if I can get a JSON response, so I can add a test case. For now - partial failure it will include the error message, for other cases it logs a default error message.